File: packages/agent/src/providers/openai.ts
// packages/agent/src/providers/openai.ts
// Dùng cho: OpenAI, OpenRouter, Ollama (tất cả OpenAI-compatible API)

import OpenAI from 'openai'
import type { SessionMessage, ToolDef, ToolCall, ToolResult } from '@quasar/core'
import { ProviderError, createLogger } from '@quasar/core'

const log = createLogger('agent:openai')

export interface CompletionOptions {
  model: string
  messages: SessionMessage[]
  tools: ToolDef[]
  systemPrompt: string
  maxTokens: number
  temperature?: number
  stream?: boolean
  onChunk?: (text: string) => void
}

export interface CompletionResult {
  content: string
  toolCalls: ToolCall[]
  usage?: { promptTokens: number; completionTokens: number; totalTokens: number }
}

export class OpenAIProvider {
  private client: OpenAI
  private providerName: string = 'OpenAI'

  constructor(apiKey: string, baseUrl?: string) {
    this.client = new OpenAI({ apiKey, baseURL: baseUrl })
    if (baseUrl?.includes('openrouter.ai')) {
      this.providerName = 'OpenRouter'
    } else if (baseUrl?.includes('ollama') || baseUrl?.includes('11434')) {
      this.providerName = 'Ollama'
    }
    log.info(`${this.providerName} provider initialized${baseUrl ? ` (${baseUrl})` : ''}`)
  }

  private buildMessages(systemPrompt: string, messages: SessionMessage[]): OpenAI.ChatCompletionMessageParam[] {
    const result: OpenAI.ChatCompletionMessageParam[] = [
      { role: 'system', content: systemPrompt },
    ]

    for (const msg of messages) {
      if (msg.role === 'user') {
        if (msg.images && msg.images.length > 0) {
          const content: OpenAI.ChatCompletionContentPart[] = [
            { type: 'text', text: msg.content || '' }
          ]
          for (const img of msg.images) {
            content.push({ type: 'image_url', image_url: { url: img } })
          }
          result.push({ role: 'user', content })
        } else {
          result.push({ role: 'user', content: msg.content })
        }
      } else if (msg.role === 'assistant') {
        const assistantMsg: OpenAI.ChatCompletionAssistantMessageParam = {
          role: 'assistant',
          content: msg.content || null,
        }
        if (msg.toolCalls && msg.toolCalls.length > 0) {
          assistantMsg.tool_calls = msg.toolCalls.map(tc => ({
            id: tc.id,
            type: 'function' as const,
            function: { name: tc.name, arguments: tc.arguments },
          }))
        }
        result.push(assistantMsg)
      } else if (msg.role === 'tool') {
        result.push({
          role: 'tool',
          tool_call_id: msg.toolCallId!,
          content: msg.content,
        })
      }
    }

    return result
  }

  private buildTools(tools: ToolDef[]): OpenAI.ChatCompletionTool[] {
    return tools.map(t => ({
      type: 'function' as const,
      function: {
        name: t.name,
        description: t.description,
        parameters: t.parameters,
      },
    }))
  }

  async complete(opts: CompletionOptions): Promise<CompletionResult> {
    const messages = this.buildMessages(opts.systemPrompt, opts.messages)
    const tools = opts.tools.length > 0 ? this.buildTools(opts.tools) : undefined

    try {
      if (opts.stream && opts.onChunk) {
        return await this.streamComplete(messages, tools, opts)
      }

      const response = await this.client.chat.completions.create({
        model: opts.model,
        messages,
        ...(tools ? { tools } : {}),
        max_tokens: opts.maxTokens,
        temperature: opts.temperature ?? 0.7,
      })

      const choice = response.choices[0]!
      const toolCalls: ToolCall[] = (choice.message.tool_calls || []).map(tc => ({
        id: tc.id,
        name: tc.function.name,
        arguments: JSON.parse(tc.function.arguments || '{}'),
      }))

      return {
        content: choice.message.content || '',
        toolCalls,
        usage: response.usage ? {
          promptTokens: response.usage.prompt_tokens,
          completionTokens: response.usage.completion_tokens,
          totalTokens: response.usage.total_tokens,
        } : undefined,
      }
    } catch (e) {
      throw new ProviderError(`${this.providerName} API error: ${e instanceof Error ? e.message : String(e)}`, e)
    }
  }

  private async streamComplete(
    messages: OpenAI.ChatCompletionMessageParam[],
    tools: OpenAI.ChatCompletionTool[] | undefined,
    opts: CompletionOptions
  ): Promise<CompletionResult> {
    const stream = await this.client.chat.completions.create({
      model: opts.model,
      messages,
      ...(tools ? { tools } : {}),
      max_tokens: opts.maxTokens,
      temperature: opts.temperature ?? 0.7,
      stream: true,
    })

    let content = ''
    const toolCallsMap = new Map<number, { id: string; name: string; args: string }>()

    for await (const chunk of stream) {
      const delta = chunk.choices[0]?.delta
      if (!delta) continue

      if (delta.content) {
        content += delta.content
        opts.onChunk?.(delta.content)
      }

      if (delta.tool_calls) {
        for (const tc of delta.tool_calls) {
          const existing = toolCallsMap.get(tc.index)
          if (existing) {
            if (tc.function?.arguments) existing.args += tc.function.arguments
          } else {
            toolCallsMap.set(tc.index, {
              id: tc.id || '',
              name: tc.function?.name || '',
              args: tc.function?.arguments || '',
            })
          }
        }
      }
    }

    const toolCalls: ToolCall[] = Array.from(toolCallsMap.values()).map(tc => ({
      id: tc.id,
      name: tc.name,
      arguments: JSON.parse(tc.args || '{}'),
    }))

    return { content, toolCalls }
  }
}

File: crates/quasar-agent/src/providers/openai.rs
use async_trait::async_trait;
use async_openai::{
    config::OpenAIConfig,
    types::{
        ChatCompletionRequestMessage, ChatCompletionRequestSystemMessage,
        ChatCompletionRequestUserMessage, ChatCompletionRequestAssistantMessage,
        ChatCompletionRequestToolMessage, ChatCompletionTool, ChatCompletionToolType,
        CreateChatCompletionRequest, CreateChatCompletionRequestArgs,
        FunctionObject, FunctionObjectArgs,
    },
    Client,
};
use quasar_core::{Message, MessageRole, QuasarResult, QuasarError};
use super::{Provider, CompletionOptions, CompletionResult, TokenUsage};

pub struct OpenAIProvider {
    client: Client<OpenAIConfig>,
}

impl OpenAIProvider {
    pub fn new(api_key: impl Into<String>, base_url: Option<impl Into<String>>) -> QuasarResult<Self> {
        let mut config = OpenAIConfig::new().with_api_key(api_key);
        
        if let Some(url) = base_url {
            config = config.with_api_base(url);
        }

        Ok(Self {
            client: Client::with_config(config),
        })
    }

    fn convert_message(&self, msg: &Message) -> ChatCompletionRequestMessage {
        match msg.role {
            MessageRole::System => {
                ChatCompletionRequestMessage::System(ChatCompletionRequestSystemMessage {
                    content: msg.content.clone().into(),
                    name: None,
                })
            }
            MessageRole::User => {
                ChatCompletionRequestMessage::User(ChatCompletionRequestUserMessage {
                    content: msg.content.clone().into(),
                    name: None,
                })
            }
            MessageRole::Assistant => {
                ChatCompletionRequestMessage::Assistant(ChatCompletionRequestAssistantMessage {
                    content: Some(msg.content.clone()),
                    tool_calls: msg.tool_calls.as_ref().map(|calls| {
                        calls
                            .iter()
                            .map(|tc| async_openai::types::ChatCompletionMessageToolCall {
                                id: tc.id.clone(),
                                r#type: ChatCompletionToolType::Function,
                                function: async_openai::types::FunctionCall {
                                    name: tc.name.clone(),
                                    arguments: tc.arguments.clone(),
                                },
                            })
                            .collect()
                    }),
                    ..Default::default()
                })
            }
            MessageRole::Tool => {
                ChatCompletionRequestMessage::Tool(ChatCompletionRequestToolMessage {
                    content: msg.content.clone().into(),
                    tool_call_id: msg.tool_call_id.clone().unwrap_or_default(),
                })
            }
        }
    }
}

#[async_trait]
impl Provider for OpenAIProvider {
    async fn complete(&self, options: CompletionOptions) -> QuasarResult<CompletionResult> {
        let mut messages = vec![self.convert_message(&Message::system(&options.system_prompt))];
        messages.extend(options.messages.iter().map(|m| self.convert_message(m)));

        let tools: Vec<ChatCompletionTool> = options
            .tools
            .iter()
            .map(|tool| {
                let function = FunctionObjectArgs::default()
                    .name(&tool.name)
                    .description(&tool.description)
                    .parameters(serde_json::to_value(&tool.parameters).unwrap_or_default())
                    .build()
                    .unwrap();

                ChatCompletionTool {
                    r#type: ChatCompletionToolType::Function,
                    function,
                }
            })
            .collect();

        let mut request_builder = CreateChatCompletionRequestArgs::default();
        request_builder
            .model(&options.model)
            .messages(messages)
            .max_tokens(options.max_tokens as u32);

        if !tools.is_empty() {
            request_builder.tools(tools);
        }

        if let Some(temp) = options.temperature {
            request_builder.temperature(temp);
        }

        let request = request_builder
            .build()
            .map_err(|e| QuasarError::provider(format!("Failed to build request: {}", e)))?;

        let response = self
            .client
            .chat()
            .create(request)
            .await
            .map_err(|e| QuasarError::provider(format!("OpenAI API error: {}", e)))?;

        let choice = response
            .choices
            .first()
            .ok_or_else(|| QuasarError::provider("No choices in response"))?;

        let content = choice
            .message
            .content
            .clone()
            .unwrap_or_default();

        let tool_calls = choice.message.tool_calls.as_ref().map(|calls| {
            calls
                .iter()
                .map(|tc| quasar_core::ToolCall {
                    id: tc.id.clone(),
                    name: tc.function.name.clone(),
                    arguments: tc.function.arguments.clone(),
                })
                .collect()
        });

        let tokens_used = TokenUsage {
            prompt_tokens: response.usage.as_ref().map(|u| u.prompt_tokens as usize).unwrap_or(0),
            completion_tokens: response.usage.as_ref().map(|u| u.completion_tokens as usize).unwrap_or(0),
            total_tokens: response.usage.as_ref().map(|u| u.total_tokens as usize).unwrap_or(0),
        };

        Ok(CompletionResult {
            content,
            tool_calls,
            tokens_used,
        })
    }
}
